[5.x] JSON value on function builder

celsowm opened the pull request:
A few years ago I opened an [issue](https://github.com/cakephp/cakephp/issues/15073) but weeks later I had some family problems.

Now I would like to try to finally turn this idea into a concrete contribution.

It's not 100% perfect yet because I don't know how to make it work in situations where the first parameter is also a string:

`SELECT JSON_VALUE( '{ "name" : "Wag", "scores" : [ 25, 36, 48 ] }', '$.scores[1]' );`

celsowm committed:
json value on function builder

celsowm commented:
Hi @ADmad , do you know how I can see files where that happened?

![image](https://github.com/cakephp/cakephp/assets/369336/5ca09cbf-59db-4b98-bd17-a6a2c0ff825b)

And do you know the reason of this error?

```
1) Cake\Test\TestCase\Mailer\Transport\SmtpTransportTest::testAuthTypeParsingIsSkippedIfNoCredentialsProvided
Error: Call to undefined method PHPUnit\Framework\MockObject\Builder\InvocationMocker::withConsecutive()
```

ADmad commented:
Please rebase your branch into upstream 5.x, that failing test case has been fixed now.

celsowm commented:
> Please rebase your branch into upstream 5.x, that failing test case has been fixed now.

@ADmad Sorry but... How can I do that?

celsowm commented:
For information only, JSON remains a focus in SQL ANSI 2023: https://peter.eisentraut.org/blog/2023/04/04/sql-2023-is-finished-here-is-whats-new

celsowm commented on src/Database/FunctionsBuilder.php:
```diff
@@ -329,6 +329,24 @@ public function lead(
+        string $jsonPath,
+        array $types = []
+    ): FunctionExpression {
+        $params = $this->toLiteralParam($expression) + [$jsonPath];
```
Is there a way to allow something like JSON_VALUE( '{ "name" : "Wag", "scores" : [ 25, 36, 48 ] }', '$.scores[1]' ); 
as result?

othercorey replied:
You want to support passing a sql literal and not an identifier as the field?

celsowm replied:
yes !

othercorey replied:
Can you do this by passing in an expression set to a string?

celsowm replied:
@othercorey How can I do this is this code?

ndm2 replied:
In situations where an API should be able to take string as well as identifiers, I'd personally prefer strings to be bound as strings, and require identifiers to be passed as (identifier) expressions.

markstory replied:
Agree with @ndm2. We can't have any methods that accept both columns names as strings, and string values for a parameter. Doing so makes it impossible for us to prevent SQL injection. If we need to accept string literal values, then column names need to be wrapped in `IdentifierExpression`. I would hope that most usage of JSON value extraction are on columns, as passing a string literal into the database so you can do a json operation on it seems wasteful.

celsowm replied:
@markstory @ndm2 You guys are right ! Thanks !

othercorey commented on tests/TestCase/Database/Query/UpdateQueryTest.php:
```diff
@@ -454,4 +457,43 @@ public function testSymmetricJsonType(): void
+        $skip = false;
+        $driver = $this->connection->getDriver();
+        $version = $this->connection->getDriver()->version();
+        if ($driver instanceof Postgres && version_compare($version, '12.0.0', '<')) {
```
Might be useful to add this to the driver feature constants.

othercorey commented on tests/TestCase/Database/FunctionsBuilderTest.php:
```diff
@@ -293,4 +293,15 @@ public function testLead(): void
+        $this->assertInstanceOf(FunctionExpression::class, $function);
+        $this->assertSame('JSON_VALUE(field, :param0)', $function->sql(new ValueBinder()));
+        $this->assertSame('string', $function->getReturnType());
+    }
```
Need to add a test passing an expression for the field.

othercorey commented:
@celsowm I did not see this earlier. The overall approach looks good. Need some more tests to confirm the cast to ::jsonb works with expressions.

celsowm commented:
Hi @othercorey ,
Are you here at cakefest 23'?

othercorey commented:
> Hi @othercorey , Are you here at cakefest 23'?

Sorry, no. I think markstory is probably more fun in person though :)

markstory commented on tests/TestCase/Database/Query/UpdateQueryTest.php:
```diff
@@ -454,4 +457,43 @@ public function testSymmetricJsonType(): void
+        } elseif ($driver instanceof Sqlite && version_compare($version, '3.19', '<')) {
+            $skip == true;
+        }
+        $this->skipIf($skip);
```
```suggestion
        $this->skipIf($skip, 'The current database backend does not support JSON value operations');
```

fabian-mcfly committed:
Mark modified fields as clean instead of the entire entity

fabian-mcfly committed:
Fixed EAV Strategy mistake & added tests

fabian-mcfly committed:
CS & Annotations

ajibarra commented:
@celsowm Hi, could you please take a look to failing check so this can be merged?

celsowm commented:
> @celsowm Hi, could you please take a look to failing check so this can be merged?

![image](https://github.com/cakephp/cakephp/assets/369336/3b445b66-dff3-494a-9243-479bb9f9d367)

Hi, sorry been late !

Do you know who can I find the file that is faling on php cs ?

othercorey commented:
> Do you know who can I find the file that is faling on php cs ?

Open the Files Changed tab and scroll down. It's just some extra whitespace.

You can also run `composer cs-check` locally.

celsowm commented:
> > Do you know who can I find the file that is faling on php cs ?
> 
> Open the Files Changed tab and scroll down. It's just some extra whitespace.
> 
> You can also run `composer cs-check` locally.

something very strange:

![image](https://github.com/cakephp/cakephp/assets/369336/e2fc067a-07b8-44c2-a6e4-ac5ce7a67139)

but if try to press "delete":

![image](https://github.com/cakephp/cakephp/assets/369336/fb861427-73e6-4cb2-8b6b-b05d0332494c)

othercorey commented:
I wonder if this is an unexpected newline causing the problem.

markstory committed:
Increase session gc lifetime

When sessions have serverside timeouts disabled we should set a longer
session.gc_maxlifetime option as using 0 could result in all sessions
being reaped when a GC sweep is done.

Refs #17513

markstory committed:
refine timeout and gclifetime code

Instead of special casing 0 with a hardcoded default, we can leave the
gc_maxlifetime alone and only define it when a timeout has been set.

fabian-mcfly committed:
CS & Annotations

dereuromark committed:
Revert for case when cell is not part of view.

dereuromark committed:
Prefer defined vars over undefined checks.

markstory committed:
Add cast

celsowm commented:
@othercorey  Finally ! All checks have passed !

othercorey commented:
@celsowm Great! Can you rebase your branch on top of 5.next so we can merge this cleanly?

celsowm commented:
> @celsowm Great! Can you rebase your branch on top of 5.next so we can merge this cleanly?

![image](https://github.com/cakephp/cakephp/assets/369336/e08cdf63-7323-44b3-8f2f-2bcbf8f08d7b)
any hints?

othercorey commented:
You need to pull 5.next and rebase on top of that then force push.

celsowm commented:
> You need to pull 5.next and rebase on top of that then force push.

When I try:
![image](https://github.com/cakephp/cakephp/assets/369336/854e9d17-0735-41e1-9af5-6e62265c779a)

the options are:
![image](https://github.com/cakephp/cakephp/assets/369336/06c9c9eb-e099-4e87-a8cb-eea1c7adb665)

othercorey commented:
On your github fork page, for 5.next, click Sync Fork. That will pull from upstream.

![image](https://github.com/cakephp/cakephp/assets/24221186/50061115-3264-4097-84bd-585953dc2063)

celsowm commented:
> On your github fork page, for 5.next, click Sync Fork. That will pull from upstream.
> 
> ![image](https://private-user-images.githubusercontent.com/24221186/296521863-50061115-3264-4097-84bd-585953dc2063.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDUyNjEzMTIsIm5iZiI6MTcwNTI2MTAxMiwicGF0aCI6Ii8yNDIyMTE4Ni8yOTY1MjE4NjMtNTAwNjExMTUtMzI2NC00MDk3LTg0YmQtNTg1OTUzZGMyMDYzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTE0VDE5MzY1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg2ZDI1YzEyOTVlNmJmN2Y0YmFhZTNmODFkOGI0MTJkMDdhMWUwNWZkOThlMzZjNmRhMzk1NzZkMmVmYzhmZjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.gcFn1FbgjgPPBLl-UjmzlwU_osVP62S7U9CfU_pkbFA)

thanks !
There were two very old commits and deleting them I were able to sync !

celsowm commented:
@othercorey now is only a type-hint thing on redis class

ADmad commented:
@celsowm Your PR still has lot of additional commits, as your feature branch hasn't been rebased on top of latest 5.next.

Try this sequence of commands:
```cli
git checkout 5.next  // if branch doesn't already exist then: git check -b 5.next upstream/5.next
git pull upstream 5.next
git checkout json-value-cake5
git rebase 5.next
git push -f origin json-value-cake5
```

othercorey commented:
I tried to sync your fork and manually rebase, but your fork doesn't seem to have the 5.x or 5.next branches.

othercorey commented:
I tried to fix this branch, but there was something incorrectly merged I think. I squashed all of the commits for this change.

Can you create a new branch off of 5.next and cherry-pick your commit to it?

celsowm commented:
> @celsowm Your PR still has lot of additional commits, as your feature branch hasn't been rebased on top of latest 5.next.
> 
> Try this sequence of commands:
> 
> ```
> git checkout 5.next  // if branch doesn't already exist then: git check -b 5.next upstream/5.next
> git pull upstream 5.next
> git checkout json-value-cake5
> git rebase 5.next
> git push -f origin json-value-cake5
> ```

![image](https://github.com/cakephp/cakephp/assets/369336/65d9bcd1-8660-4294-8bb8-15a16f023368)

celsowm committed:
Merge branch '5.next' into json-value-cake5

markstory commented:
Merged in 5.next in b54352e

ADmad commented:
@celsowm `git remote add upstream git@github.com:cakephp/cakephp.git` to add a new remote repo named "upstream" and then use the commands I gave. (The "check" instead of "checkout" was a typo).

Learning git CLI instead of fiddling with GH UI or IDE buttons serves better in the long run :)
